### PR TITLE
VM metrics

### PR DESCRIPTION
**Pull request by bmacnaughton**
N.B. requires bindings v9 (implements `Reporter.sendMetrics()`) to be released before this.

This release implements [AO-10912] and minor test updates.

- implement `sendMetrics()` which calls `Reporter.sendMetrics()`.
- rewrite `sendMetric()` as a wrapper around `sendMetrics()` so non-breaking
- add `sendMetrics()` to guides/api.md
- deprecate `sendMetric()` in guides/api.md
- config file `runtimeMetrics` or env var `APPOPTICS_RUNTIME_METRICS`

[AO-10912]: https://swicloud.atlassian.net/browse/AO-10912

### CODE REVIEW

File: lib/metrics.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  addMetricV (metric, value, n = 1) {
+    this.metrics.push({
+      name: `${this.prefix}.${metric}`,
+      count: n,
+      addHostTag: true,
+      value,
+    });
+  }
+
+  addMetricI (metric, n = 1) {
```
Comment (prettycoder) [inline lib/metrics.js]: can't find addMetricI being used anywhere

File: test/versions.js [hunk 2]
```diff
@@ -98,7 +103,7 @@ test('mysql', '>= 2.1.0')
 test('oracledb', '>= 2.0.14')
 
 test('pino', '>= 2.3.0');
-test('pg', '>= 4.5.5')
+test('pg', '>= 4.5.5 < 7.0.0 || >= 7.5.0')
```
[the remark this answers is not in the document]
Comment (bmacnaughton) [reply test/versions.js]: the versions being excluded fail the tests. it's not clear that the instrumentation will fail. i spent way too much time tracking down the changes that `pg` made. in 7.0.0 they changed the way errors were reported which is what causes the test failure; it wasn't real error but they started reporting the error synchronously. in 7.5.0 they started deferring the error reporting with process.nextTick() which allows the test to complete successfully.

i have already updated the support matrix locally but did not put the gap into it. if we run into a customer with an issue i'll revisit it. but the current version of `pg` is 23+ versions beyond the gap most people are going to be past it.

File: lib/get-unified-config.js [hunk 3]
```diff
@@ -37,6 +38,7 @@ function getUnifiedConfig () {
     hostnameAlias: {location: 'b', type: 's'},
     ec2MetadataTimeout: {location: 'b', type: 'i'},
     triggerTraceEnabled: {location: 'b', type: 'b', default: true},
+    runtimeMetrics: {location: 'b', type: 'b', default: true},
```
[the remark this answers is not in the document]
Comment (bmacnaughton) [reply lib/get-unified-config.js]: yes, it should have been. and i added updating configuration.md to my nascent checklist. thanks.

File: lib/api-sim.js [hunk 4]
```diff
@@ -35,11 +35,19 @@ module.exports = function (appoptics) {
     return span
   }
 
+  class Metrics {
+    constructor () {}
+    start () {}
+    stop () {}
+    resetInterval () {}
```
[the remark this answers is not in the document]
Comment (bmacnaughton) [reply lib/api-sim.js]: the criteria for the functions here is that 1) they are called in the normal course of operations and 2) they existed when i filled this in. i added getStatus() later but never referenced it. it was an afterthought to provide an alternative to just looking at the status property directly. maybe i should get rid of it.

File: lib/metrics.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    this.metrics.push({
+      name: `${this.prefix}.${metric}`,
```
[the remark this answers is not in the document]
Comment (bmacnaughton) [reply lib/metrics.js]: ok

File: lib/metrics.js (hunk 1 above)
Comment (bmacnaughton) [reply lib/metrics.js]: nope, it's not. but it's there for completeness so i don't have to go look it up if i end up needing it.

File: lib/api.js [hunk 6]
```diff
@@ -1166,7 +1171,112 @@ function reportInfo (data) {
...
+  const merged = [];
+  for (let i = 0; i < metrics.length; i++) {
+    const m = metrics[i];
+
+    const metric = Object.assign({addHostTag}, m);
+    metric.tags = Object.assign({}, globalTags, m.tags);
+    merged.push(metric);
+  }
+
+  return aob.reporter.sendMetrics(merged);
```
[the remark this answers is not in the document]
Comment (bmacnaughton) [reply lib/api.js]: yes, it should. this one if giving me a bit of trouble right now - `prepack.js` runs before packing and/or publishing and, among other things, verifies that api and api-sim are in sync. but it's having trouble running right now - related to metrics. more to come.

**Comment by bmacnaughton** [comment]
> Maybe a suite that fails if bindings and this code are out of sync.
in some respects every test that is run implicitly tests this. is there something specific that you're thinking of?

File: lib/api.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (bmacnaughton) [reply lib/api.js]: the circular dependency resolution was exercised by `prepack.js` in a different order than it usually is. as a result the new class `Metrics` was not resolved when `index.js` tried to invoke the class. the solution i chose was to move starting metrics into the same deferred function as is used to send the init message. that function is called by `process.nextTick()` which allows resolution of the circular dependency.

**Pull request merged**